### DISCUSSION TITLE
integrate community contributed API layer into Caido

### DISCUSSION THREAD

**Discussion by Riddle1001** [caido/caido#1619] [Q&A]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help us know what to work on
- Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request

### Description
Implement a integrated secondary API layer within Caido that consists of community contributed functions. This layer would be built directly into Caido. It allows for continuous community driven development. This secondary API would be put into Caido by default and not require any additional plugin, meaning any plugin is able to access this API.

For example, we could have custom UI components , custom events, onPageSelected, onCaidoLoad, a helper function to parse and return a HTTP request from a raw HTTP string all created by the community.

**Comment by Sytten** [comment]
I am not sure we can ship this API natively, because that means we would have to be very involved to make sure of security and backward compatibility. But we can make it very easy to install and/or have a dependency requirement system that tells the user to install it.

**Comment by Riddle1001** [comment]
> I am not sure we can ship this API natively, because that means we would have to be very involved to make sure of security and backward compatibility. But we can make it very easy to install and/or have a dependency requirement system that tells the user to install it.

I understand the security part, but ideally, this second layer API would be community driven. While evenbetter is doing a good job, and hopefully so will other APIs, there still remains the possibility of someone creating a API for developers to use, and many plugins using it, and then the API developer stops maintaining the API, taking those plugins down with it. The community driven aspect of this second layer API is to hopefully have a never dying, almost always backward compatible API due to the open source nature of it, and since Caido isn't going to leave Caido, there would be no issues of people using it and then it being abandoned. I think overly complex functions should not be wanted, but simple stuff like creating event listeners for tab changes, Caido init, parsing a raw HTTP request into an object, etc, this would help some with the security aspect in terms.

**Comment by Corb3nik** [comment]
This is available via https://github.com/caido-community/utilities now :) A third-party library with additional utilities for Caido
